blackbird_python/blackbird/program.py

```diff
@@ -0,0 +1,290 @@
+                        var_name = "A{}".format(var_count)
+                        args.append(var_name)
+                        var_count += 1
+
+                        # add array declaration to script after the metadata block
+                        bb_array = numpy_to_blackbird(v, var_name)
+                        for idx, line in enumerate(bb_array):
+                            script.insert(array_insert + idx, line)
+
+                        array_insert += len(bb_array)
+
+                    elif isinstance(v, str):
+                        # argument is a string type
+                        args.append('"{}"'.format(v))
+
+                    elif isinstance(v, complex):
+                        # argument is a complex type
+                        args.append(
+                            "{}{}{}j".format(v.real, "+-"[int(v.imag < 0)], np.abs(v.imag))
+                        )
+
+                    else:
+                        args.append("{}".format(v))
+
+                # loop through keyword argument
+                for k, v in op["kwargs"].items():
+                    # for each operation argument, format it
+                    # correctly depending on its type
+                    if isinstance(v, np.ndarray):
+                        # create an array variable
+                        var_name = "A{}".format(var_count)
+                        kwargs.append('{}={}'.format(k, var_name))
+                        var_count += 1
+
+                        # add array declaration to script
+                        bb_array = numpy_to_blackbird(v, var_name)
+                        for idx, line in enumerate(bb_array):
+                            script.insert(array_insert + idx, line)
+
+                        array_insert += len(bb_array)
+
+                    elif isinstance(v, str):
+                        kwargs.append('{}="{}"'.format(k, v))
+
+                    elif isinstance(v, complex):
+                        kwargs.append(
+                            "{}={}{}{}j".format(k, v.real, "+-"[int(v.imag < 0)], np.abs(v.imag))
+                        )
+
+                    else:
+                        kwargs.append("{}={}".format(k, v))
+
+                if args and kwargs:
+                    # arguments = "({}, {})".format(", ".join(args), ", ".join(kwargs))
+                    raise ValueError(
+                        "Blackbird currently doesn't support operations "
+                        "which contain both positional and keyword arguments. "
+                        "\nPlease provide all operation arguments as keyword arguments for now."
+                    )
+                elif not kwargs:
+                    arguments = "({})".format(", ".join(args))
+                elif not args:
+                    arguments = "({})".format(", ".join(kwargs))
+
+                script.append("{}{} | {}".format(op["op"], arguments, modes))
+            else:
```

Comment:
this occurs far down from the original conditional statement. Would be helpful to put a comment stating what is happening with `else`